microprofile/opentracing-smallrye/src/main/java/org/wildfly/microprofile/opentracing/smallrye/TracingCDIExtension.java

```diff
@@ -19,26 +19,45 @@
  * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
  * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
  */
-
 package org.wildfly.microprofile.opentracing.smallrye;
 
 import io.opentracing.Tracer;
+import java.util.Collections;
+import java.util.Map;
+import java.util.WeakHashMap;
 
 import javax.enterprise.event.Observes;
-import javax.enterprise.inject.spi.BeanManager;
-import javax.enterprise.inject.spi.BeforeBeanDiscovery;
+import javax.enterprise.inject.spi.AfterBeanDiscovery;
+import javax.enterprise.inject.spi.BeforeShutdown;
 import javax.enterprise.inject.spi.Extension;
 import javax.enterprise.inject.spi.ProcessAnnotatedType;
 
 public class TracingCDIExtension implements Extension {
 
-    public void observeBeforeBeanDiscovery(@Observes BeforeBeanDiscovery bbd, BeanManager manager) {
-        String extensionName = TracingCDIExtension.class.getName();
-        bbd.addAnnotatedType(manager.createAnnotatedType(TracerProducer.class), extensionName + "-" + TracerProducer.class.getName());
+    private static final Map<ClassLoader, Tracer> TRACERS = Collections.synchronizedMap(new WeakHashMap<>());
+
+    public static void registerApplicationTracer(ClassLoader classLoader, Tracer tracerInstance) {
+        TRACERS.put(classLoader, tracerInstance);
+    }
+
+    public void registerTracerBean(@Observes AfterBeanDiscovery abd) {
+        abd.addBean().addTransitiveTypeClosure(Tracer.class).produceWith(i -> {
```

Comment:
Maybe also add `.scope(ApplicationScoped.class)`. This won't have any functional impact (apart from using proxy) as you are already sharing the instance, but it is a "cleaner" way to declare a bean that in fact is app-wide.

Reply:
@ehsavoie Did this input get dropped?

Reply:
We talked on zulip. Apparently the proxy interferes with some places where the contextual instance is needed.

Reply:
I'm currently trying to fix the test suite which is talking more time then I thought

Reply:
This change breaks with ear as the tracer will now be shared across all the ear and so all its wars